R/kwic.R

```diff
@@ -88,40 +88,43 @@ kwic.corpus <- function(x, pattern, window = 5,
 #'          "Is it a bird or a plane or is it a train?")
 #' kwic(txt, c("is", "a"), valuetype = "fixed")
 #' kwic(txt, phrase(c("is", "a", "is it")), valuetype = "fixed")
-#' 
+#'
 #' toks <- tokens(txt)
 #' kwic(toks, c("is", "a"), valuetype = "fixed")
 #' kwic(toks, phrase(c("is", "a", "is it")), valuetype = "fixed")
-#' 
+#'
 #' corp <- corpus(txt)
 #' kwic(corp, c("is", "a"), valuetype = "fixed", separator = "", remove_separators = FALSE)
-#' @export 
-kwic.tokens <- function(x, pattern, window = 5, 
-                        valuetype = c("glob", "regex", "fixed"), 
-                        separator = " ", 
+#' @export
+kwic.tokens <- function(x, pattern, window = 5,
+                        valuetype = c("glob", "regex", "fixed"),
+                        separator = " ",
                         case_insensitive = TRUE, ...) {
-    
+
     valuetype <- match.arg(valuetype)
     types <- types(x)
-    
+
     # add document names if none
     if (is.null(names(x))) {
         names(x) <- paste0(quanteda_options("base_docname"), seq_len(x))
     }
-    
-    keywords_id <- pattern2list(pattern, types, 
-                              valuetype, case_insensitive, attr(x, 'concatenator'))
+
+    keywords_id <- pattern2list(pattern, types,
+                              valuetype, case_insensitive, attr(x, "concatenator"))
     temp <- qatd_cpp_kwic(x, types, keywords_id, window, separator)
-    
+
     # attributes for kwic object
-    result <- structure(temp, 
-                        class = c("kwic", "data.frame"), 
-                        ntoken = ntoken(x), 
-                        valuetype = valuetype, 
-                        keywords = attr(keywords_id, 'pattern'),
-                        tokens =  attr(temp, "tokens"))
+    result <- structure(temp,
+                        class = c("kwic", "data.frame"),
+                        ntoken = ntoken(x),
+                        valuetype = valuetype,
+                        keywords = if (is.dictionary(pattern)) {
```

Comment:
@koheiw Here is the key change, you might know a better way to do this.

Reply:
I am not sure if this is the cause 
```r
quanteda:::pattern2list(dictionary(list(A = "a*", B = "b*")), 
                        c("apple", "apricot", "banana"), "glob", TRUE)
```
retunrs
```r
$A
[1] 1

$A
[1] 2

$B
[1] 3

attr(,"key")
[1] "A" "B"
```
so `unique(names(keywords_id))` should be enough.

Reply:
If you want dictionary values instead of keys, why don't you unlist before `pattern2list()`?

Reply:
I'm pretty sure I get what you're saying, but since you're the grand master of the `pattern2list` mojo, how about you modify that part of this PR (if you think it needs it)?